pennylane/beta/qnode.py

```diff
@@ -0,0 +1,540 @@
+            strings are allowed:
+
+            * ``"best"``: Best available method. Uses classical backpropagation or the
+              device directly to compute the gradient if supported, otherwise will use
+              the analytic parameter-shift rule where possible with finite-difference as a fallback.
+
+            * ``"device"``: Queries the device directly for the gradient.
+              Only allowed on devices that provide their own gradient computation.
+
+            * ``"backprop"``: Use classical backpropagation. Only allowed on simulator
+              devices that are classically end-to-end differentiable, for example
+              :class:`default.tensor.tf <~.DefaultTensorTF>`. Note that the returned
+              QNode can only be used with the machine-learning framework supported
+              by the device.
+
+            * ``"adjoint"``: Uses an `adjoint method <https://arxiv.org/abs/2009.02823>`__ that
+              reverses through the circuit after a forward pass by iteratively applying the inverse
+              (adjoint) gate. Only allowed on supported simulator devices such as
+              :class:`default.qubit <~.DefaultQubit>`.
+
+            * ``"parameter-shift"``: Use the analytic parameter-shift
+              rule for all supported quantum operation arguments, with finite-difference
+              as a fallback.
+
+            * ``"finite-diff"``: Uses numerical finite-differences for all quantum operation
+              arguments.
+
+            * ``None``: QNode cannot be differentiated. Works the same as ``interface=None``.
+
+        max_expansion (int): The number of times the internal circuit should be expanded when
+            executed on a device. Expansion occurs when an operation or measurement is not
+            supported, and results in a gate decomposition. If any operations in the decomposition
+            remain unsupported by the device, another expansion occurs.
+        mode (str): Whether the gradients should be computed on the forward
+            pass (``forward``) or the backward pass (``backward``). Only applies
+            if the device is queried for the gradient; gradient transform
+            functions available in ``qml.gradients`` are only supported on the backward
+            pass.
+        cache (bool or dict or Cache): Whether to cache evaluations. This can result in
+            a significant reduction in quantum evaluations during gradient computations.
+            If ``True``, a cache with corresponding ``cachesize`` is created for each batch
+            execution. If ``False``, no caching is used. You may also pass your own cache
+            to be used; this can be any object that implements the special methods
+            ``__getitem__()``, ``__setitem__()``, and ``__delitem__()``, such as a dictionary.
+        cachesize (int): The size of any auto-created caches. Only applies when ``cache=True``.

```

Comment:
Uhhh yes I think so 🙂

Reply:
Not needed for this PR, but at some point we should include an example with passing a custom cache object because I don't think I could figure out just from the docstrings how it would work.

Reply:
Oh, good idea! It ends up being as simple as doing

```python
my_cache= {}
@qml.beta.qnode(dev, cache=my_cache)
def circuit():
```
since dictionaries are valid caches. But you could also do something more complicated like
```python
from cachetools import LRUCache

my_cache = LRUCache(cachesize=1000)
@qml.beta.qnode(dev, cache=my_cache)
def circuit():
```

Reply:
Oh gotcha! Okay that's pretty straightforward :)

Reply:
I was curious after leaving that comment, so decided to play around with it:

![image](https://user-images.githubusercontent.com/2959003/134006370-18aa2782-22bc-4248-a4da-def5d7396704.png)

Reply:
Here the `key` corresponds to `tape.hash`.